### ISSUE TITLE
Bug: score of zero for function_score with nested bool filter query

### ISSUE THREAD

**Issue by lukas-gitl** [elastic/elasticsearch#27419]
>_We are currently blocked by this. I'd much appreciate some feedback._ 

**Problem Description**:

Score of zero returned when using `function_score` with query that has a `nested bool filter`. Expected is that the function score actually reflects the `random_score` function result.

Using `explain` it seems that the query pollutes the score. According to [the documentation](https://www.elastic.co/guide/en/elasticsearch/reference/5.4/query-dsl-function-score-query.html#query-dsl-function-score-query) this should never happen when `boost_mode: replace` is used. Removing the nesting also fixes this.

[This guy](https://discuss.elastic.co/t/range-filters-with-query-string-and-function-score-returning-zero-results/98472) seems to have the same problem. Unfortunately they never received an answer.

**Offending query:**
```json
{
  "query": {
    "function_score": {
      "query": {
        "nested": {
          "path": "schedules",
          "query": {
            "bool": {
              "filter": [
                {
                  "range": {
                    "schedules.start": {
                      "lte": "now"
                    }
                  }
                }
              ]
            }
          }
        }
      },
      "random_score": {},
      "score_mode": "sum",
      "boost_mode": "replace"
    }
  },
  "explain": true
}
```

**Steps to reproduce**:

Download [data.zip](https://github.com/elastic/elasticsearch/files/1480690/data.zip), unzipped it into a folder, cd into the folder and then run `. run.sh`

**Debug Output**:
```
{
  "took" : 3,
  "timed_out" : false,
  "_shards" : {
    "total" : 5,
    "successful" : 5,
    "failed" : 0
  },
  "hits" : {
    "total" : 1,
    "max_score" : 0.0,
    "hits" : [
      {
        "_shard" : "[item][2]",
        "_node" : "Lbt3Gxq7RnSir3byoNpp5w",
        "_index" : "item",
        "_type" : "item",
        "_id" : "7517967e-824e-4bfd-b67c-887cfeeeabb8",
        "_score" : 0.0,
        "_source" : {
          "id" : "7517967e-824e-4bfd-b67c-887cfeeeabb8",
          "schedules" : {
            "start" : "2015-01-01"
          }
        },
        "_explanation" : {
          "value" : 0.0,
          "description" : "sum of:",
          "details" : [
            {
              "value" : 0.0,
              "description" : "function score, product of:",
              "details" : [
                {
                  "value" : 0.0,
                  "description" : "Score based on 1 child docs in range from 0 to 0, best match:",
                  "details" : [
                    {
                      "value" : 0.0,
                      "description" : "sum of:",
                      "details" : [
                        {
                          "value" : 0.0,
                          "description" : "ConstantScore(schedules.start:[-9223372036854775808 TO 9223372036854775807]), product of:",
                          "details" : [
                            {
                              "value" : 0.0,
                              "description" : "boost",
                              "details" : [ ]
                            },
                            {
                              "value" : 1.0,
                              "description" : "queryNorm",
                              "details" : [ ]
                            }
                          ]
                        },
                        {
                          "value" : 0.0,
                          "description" : "match on required clause, product of:",
                          "details" : [
                            {
                              "value" : 0.0,
                              "description" : "# clause",
                              "details" : [ ]
                            },
                            {
                              "value" : 1.0,
                              "description" : "_type:__schedules, product of:",
                              "details" : [
                                {
                                  "value" : 1.0,
                                  "description" : "boost",
                                  "details" : [ ]
                                },
                                {
                                  "value" : 1.0,
                                  "description" : "queryNorm",
                                  "details" : [ ]
                                }
                              ]
                            }
                          ]
                        }
                      ]
                    }
                  ]
                },
                {
                  "value" : 0.6308505,
                  "description" : "min of:",
                  "details" : [
                    {
                      "value" : 0.6308505,
                      "description" : "random score function (seed: -949337162)",
                      "details" : [ ]
                    },
                    {
                      "value" : 3.4028235E38,
                      "description" : "maxBoost",
                      "details" : [ ]
                    }
                  ]
                }
              ]
            },
            {
              "value" : 0.0,
              "description" : "match on required clause, product of:",
              "details" : [
                {
                  "value" : 0.0,
                  "description" : "# clause",
                  "details" : [ ]
                },
                {
                  "value" : 1.0,
                  "description" : "#*:* -_type:__*, product of:",
                  "details" : [
                    {
                      "value" : 1.0,
                      "description" : "boost",
                      "details" : [ ]
                    },
                    {
                      "value" : 1.0,
                      "description" : "queryNorm",
                      "details" : [ ]
                    }
                  ]
                }
              ]
            }
          ]
        }
      }
    ]
  }
}
```

----------

**Elasticsearch version**:

```json
"number" : "5.1.2",
"build_hash" : "c8c4c16",
"build_date" : "2017-01-11T20:18:39.146Z",
"build_snapshot" : false,
"lucene_version" : "6.3.0"
```

**Plugins installed**: -

**JVM version**:
```
java version "1.8.0_151"
Java(TM) SE Runtime Environment (build 1.8.0_151-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.151-b12, mixed mode)
```
**OS version**:
```
Ubuntu 16.04
Linux fluxbox 4.10.0-38-generic #42~16.04.1-Ubuntu SMP 
Tue Oct 10 16:32:20 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by colings86** [comment]
@martijnvg could you take a look at this?

**Comment by jimczi** [comment]
This bug was fixed in 5.2.0:
https://github.com/elastic/elasticsearch/pull/22172
The inner range query is rewritten on the shards but the custom `boost_mode` is lost in this operation and the default value (`multiply`) is used. 
I tested with 5.6 and it works as expected, please upgrade to a newer version.

**Comment by lukas-gitl** [comment]
@jimczi Sweet. Thank you very much for the quick response!
